### ISSUE TITLE
callback function types are missing `extern(C)`

### ISSUE THREAD

**Issue by huglovefan** [BindBC/bindbc-lua#11]
the callback types like `lua_CFunction` are currently declared without an `extern()` so they default to `extern(D)`, but the [calling convention](https://en.wikipedia.org/wiki/Calling_convention) for those isn't compatible with `extern(C)` in all cases:

- according to [issue 20204], `extern(D)` functions are passed their arguments in reverse order compared to `extern(C)` (this doesn't affect the common `lua_CFunction` since it has just 1 parameter)
- windows/x86 has its D calling convention documented [here], which is supposedly different enough from the C one that it had to be specifically documented

[issue 20204]: https://issues.dlang.org/show_bug.cgi?id=20204
[here]: https://dlang.org/spec/abi.html#function_calling_conventions

one of the types this affects is custom allocator functions since they have 4 parameters:

```D
/+ dub.sdl:
dependency "bindbc-lua" version="~>0.4.1"
subConfigurations "bindbc-lua" "static"  platform="posix"
versions "BindLua_Static" "LUA_54"       platform="posix"
libs "lua5.4"                            platform="posix"
+/
import core.stdc.stdlib;
import std.exception;
import std.stdio;
import bindbc.lua;

// https://www.lua.org/manual/5.4/manual.html#lua_Alloc
extern (C) // <-- change to extern(D) and it breaks
void* myalloc(void* ud, void* ptr, size_t osize, size_t nsize) nothrow
{
	writefln("ud=%s ptr=%s osize=%s nsize=%s", ud, ptr, osize, nsize).assumeWontThrow;
	if (nsize == 0)
		{ free(ptr); return null; }
	else
		return realloc(ptr, nsize);
}

// lua_Alloc with D calling convention (same as bindbc-lua)
alias lua_Alloc_D = extern (D) void* function(void* ud, void* ptr, size_t osize, size_t nsize) nothrow;

void main()
{
	lua_State* L = luaL_newstate();
	lua_setallocf(L, cast(lua_Alloc_D)&myalloc, null);
	luaL_openlibs(L);
	if (LUA_OK != luaL_dostring(L, `
	print("hello world")
	`)) lua_error(L);
	lua_close(L);
}
```

with `extern(C)` on the allocator function, it works normally but if changed to `extern(D)`, it prints bogus values and crashes (tested on linux with dmd, 64-bit and 32-bit)

in general, any function that's going to be called by C code should have `extern(C)` to work correctly

**Comment by mdparker** [comment]
Totally an oversight. Callbacks in C libraries should always be `extern(C)`. I'm surprised it hasn't come up before. I'll fix it soon.

**Comment by mdparker** [comment]
This is fixed in 0.5.0. Thanks!
